linkerd/http/retry/src/compat.rs

```diff
@@ -0,0 +1,115 @@
+//! Compatibility utilities for upgrading to http-body 1.0.
+
+use http_body::Body;
+
+pub(crate) use self::frame::Frame;
+
+mod frame;
+
+#[derive(Debug)]
+pub(crate) struct ForwardCompatibleBody<B> {
+    inner: B,
+    data_finished: bool,
+    trailers_finished: bool,
+}
+
+// === impl ForwardCompatibleBody ===
+
+impl<B: Body> ForwardCompatibleBody<B> {
+    pub(crate) fn new(body: B) -> Self {
+        if body.is_end_stream() {
+            Self {
+                inner: body,
+                data_finished: true,
+                trailers_finished: true,
+            }
+        } else {
+            Self {
+                inner: body,
+                data_finished: false,
+                trailers_finished: false,
+            }
+        }
+    }
+
+    pub(crate) fn into_inner(self) -> B {
+        self.inner
+    }
+
+    /// Returns a future that resolves to the next frame.
+    pub(crate) fn frame(&mut self) -> combinators::Frame<'_, B> {
+        combinators::Frame(self)
+    }
+}
+
+/// Future that resolves to the next frame from a `Body`.
+///
+/// NB: This is a vendored stand-in for [`Frame<'a, T>`][frame], and and can be replaced once
+/// we upgrade from http-body 0.4 to 1.0. This file was vendored, and subsequently adapted to this
+/// project, at commit 86fdf00.
```

Comment:
"vendored" here is doing some heavy lifting, as the `poll()` implementation must do a bit more work to determine whether we should delegate down to `poll_data()` or `poll_data()`, but the "shape" of this in terms of its `Future::Output` etc., are the same.